### PR TITLE
Complete wrapped functions

### PR DESCRIPTION
**Pull request by xingzhongyu**
add celltype annotation valid. 
Generate a subset configuration file for the pipeline from step 2.
Generate result table based on sweep_id.
Generate the configuration file of step 3 based on the result table of step 2.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -13,3 +13,5 @@ wandb
 *.egg*
 __pycache__
 build/
+
+data
```
Comment (RemyLau) [inline .gitignore:17]: This needs to be removed. There's a `dance/data` module. Ignoring `data` will cause problems tracking changes in the `dance.data` module

File: dance/pipeline.py [hunk 2]
```diff
@@ -856,7 +857,58 @@ def flatten_dict(d, *, parent_key="", sep="_"):
...
+
+    # Combine required elements with optional combinations to get all possible combinations
+    all_combinations = []
+    for optional_combination in optional_combinations:
+        all_combinations.append([x for x in elements if x in required_elements or x in optional_combination])
+    return all_combinations
+
+
+def generate_subsets(path, tune_mode, save_directory, required_indexes):
+    """Generated automatically subsets for original yaml. YAML can be generated
```
Comment (RemyLau) [inline dance/pipeline.py:886]: This is very unclear. Perhaps you can give an example besides the verbal description. I can help improve the descriptions later.

File: dance/transforms/normalize.py [hunk 3]
```diff
@@ -206,11 +206,11 @@ def __init__(
...
         split_names: Optional[Union[Literal["ALL"], List[str]]] = None,
         batch_key: Optional[str] = None,
-        min_cells: int = 5,
+        min_cells: int = 2,
         gmean_eps: int = 1,
-        n_genes: Optional[int] = 2000,
+        n_genes: Optional[int] = 3000,
         n_cells: Optional[int] = None,
-        bin_size: int = 500,
+        bin_size: int = 200,
```
Comment (RemyLau) [inline dance/transforms/normalize.py:213]: Are the purpose of these changes to the default setting to be consistent with the original implementations? In practice, we must avoid changes like this as it easily breaks user compatibility in the sense that whenever they upgrade the dance version, artifacts like these will "silently" change their results.

For now this is fine, because we have not released this in the latest version yet. But please keep in mind that once this is released in a stable version, we should be very careful about updating default options. If we have to, we must give explicit user warnings.

File: dance/transforms/normalize.py [hunk 4]
```diff
@@ -250,15 +250,14 @@ def _get_idx_dict(self, data) -> List[Dict[str, List[int]]]:
...
+    def __call__(self, data: Data):
         if isinstance(data.data.X, sp.spmatrix):
             self.logger.warning("Native support for sparse matrix is not implemented yet, "
                                 "converting to dense array explicitly.")
             data.data.X = data.data.X.A
-
         # idx_dict = self._get_idx_dict(data)
         # for name, idx in idx_dict.items():
-        selected_data = data.data.copy()
+        selected_data = data.data
```
Comment (RemyLau) [inline dance/transforms/normalize.py:260]: Can we add an option for deciding whether to overwrite the original `X` matrix or saving it as a new layer? It might be good to have the option to keep the original `X` and save the normalized matrix to `adata.layers[self.out]`.

File: examples/tuning/cta_svm/config_yamls/subset_0_pipeline_tuning_config.yaml
Comment (RemyLau) [inline examples/tuning/cta_svm/config_yamls/subset_0_pipeline_tuning_config.yaml:1]: What are these sub-configs for?

File: examples/tuning/cta_svm/main.py [hunk 5]
```diff
@@ -100,3 +106,21 @@ def evaluate_pipeline():
 $ python main.py --tune_mode (pipeline/params) --species mouse --tissue Kidney --train_dataset 4682 --test_dataset 203
 
 """
+"""
+The 10 processes with the highest test_acc in step 2
```
Comment (RemyLau) [inline examples/tuning/cta_svm/main.py:110]: If you were to put tables like this in the source code, please make it readable. For example, show less decimal points (for the sake of an example, 3 decimal points should be enough) and align the columns.

**Review by RemyLau — CHANGES_REQUESTED** [review]

File: dance/transforms/normalize.py (hunk 4 above)
Comment (xingzhongyu) [reply dance/transforms/normalize.py:260]: I'm not sure because there is saveRaw function that helps save raw data

File: dance/pipeline.py [hunk 6]
```diff
@@ -856,7 +857,60 @@ def flatten_dict(d, *, parent_key="", sep="_"):
...
+    automatically, but obviously still need to manually tune the parameters to avoid
+    errors. When part of the process is omitted, the function parameters need to be
+    changed, otherwise an error will be reported, so different YAML adjustments are
+    required.
+
+    Conveniently generate multiple yamls, such as generating (A, B, C), (A, B), (A, C), (B, C), (A), (B), (C) yaml based on pipelines A, B, and C for the next experiment
```
Comment (RemyLau) [inline dance/pipeline.py:892]: Do you think it make sense to add an option to the pipeline planner so that it automatically generate subsets like this?

File: dance/metadata/scdeepsort.csv [hunk 7]
```diff
@@ -62,7 +62,9 @@ human,Spleen,4029,train,train_human_Spleen4029_celltype.csv,https://dl.dropboxus
...
-mouse,Brain,3285,train,train_mouse_Brain3285_celltype.csv,https://www.dropbox.com/s/ozsobozk3ihkrqg?dl=1,train_mouse_Brain3285_data.csv,https://www.dropbox.com/s/zjrloejx8iqdqsa?dl=1
+mouse,Brain,3285,valid,valid_mouse_Brain3285_celltype.csv,https://www.dropbox.com/s/ozsobozk3ihkrqg?dl=1,valid_mouse_Brain3285_data.csv,https://www.dropbox.com/s/zjrloejx8iqdqsa?dl=1
 mouse,Brain,753,train,train_mouse_Brain753_celltype.csv,https://www.dropbox.com/s/x2katwk93z06sgw?dl=1,train_mouse_Brain753_data.csv,https://www.dropbox.com/s/3f3wbplgo3xa4ww?dl=1
 mouse,Kidney,4682,train,train_mouse_Kidney4682_celltype.csv,https://www.dropbox.com/s/3plrve7g9v428ec?dl=1,train_mouse_Kidney4682_data.csv,https://www.dropbox.com/s/olf5nirtieu1ikq?dl=1
 mouse,Spleen,1970,train,train_mouse_Spleen1970_celltype.csv,https://www.dropbox.com/s/3ea64vk546fjxvr?dl=1,train_mouse_Spleen1970_data.csv,https://www.dropbox.com/s/c4te0fr1qicqki8?dl=1
+mouse,Kidney,4682,valid,valid_mouse_Kidney4682_celltype.csv,https://www.dropbox.com/s/3plrve7g9v428ec?dl=1,valid_mouse_Kidney4682_data.csv,https://www.dropbox.com/s/olf5nirtieu1ikq?dl=1
```
Comment (xingzhongyu) [inline dance/metadata/scdeepsort.csv:69]: add valid dataset

File: dance/pipeline.py (hunk 6 above)
Comment (xingzhongyu) [reply dance/pipeline.py:892]: It makes sense to me. If I don't use this method, I need to copy and paste a lot of yaml and then modify it. For users, I am not sure whether this method is needed.

File: examples/tuning/cta_svm/main.py (hunk 5 above)
Comment (xingzhongyu) [reply examples/tuning/cta_svm/main.py]: changed

File: .gitignore (hunk 1 above)
Comment (xingzhongyu) [reply .gitignore]: changed

File: examples/tuning/cta_svm/results/params/1st_best_test_acc.csv [hunk 8]
```diff
@@ -0,0 +1,2 @@
+id,_step,_wandb_runtime,_runtime,test_acc,_timestamp,acc,params.0.order,params.0.max_cells,params.0.min_cells,params.0.max_counts,params.0.min_counts,params.1.n_components
```
Comment (xingzhongyu) [inline examples/tuning/cta_svm/results/params/1st_best_test_acc.csv:1]: get result by save_summary_data

File: examples/tuning/cta_svm/pipline_run.sh [hunk 9]
```diff
@@ -0,0 +1,8 @@
+# python examples/tuning/cta_svm/main.py --config_dir=config_yamls/subset_0_  # don't need 4 7 4 28 16 28
```
Comment (xingzhongyu) [inline examples/tuning/cta_svm/pipline_run.sh:1]: run all subsets to get comprehensive results.

File: examples/tuning/cta_svm/config_yamls/1st_test_acc_params_tuning_config.yaml [hunk 10]
```diff
@@ -0,0 +1,65 @@
+type: preprocessor
```
Comment (xingzhongyu) [inline examples/tuning/cta_svm/config_yamls/1st_test_acc_params_tuning_config.yaml:1]: The test_acc of step 2 obtains the highest score of yaml in the parameter adjustment of step 3.

File: examples/tuning/cta_svm/config_yamls/2nd_test_acc_params_tuning_config.yaml [hunk 11]
```diff
@@ -0,0 +1,65 @@
+type: preprocessor
```
Comment (xingzhongyu) [inline examples/tuning/cta_svm/config_yamls/2nd_test_acc_params_tuning_config.yaml:1]: Under the second highest configuration of test_acc in step2, the parameter range in step3

File: examples/tuning/cta_svm/pipeline_tuning_config.yaml [hunk 12]
```diff
@@ -28,7 +54,7 @@ pipeline:
 wandb:
   entity: danceteam
   project: dance-dev
-  method: bayes
+  method: grid #try grid to provide a comprehensive search
```
Comment (xingzhongyu) [inline examples/tuning/cta_svm/pipeline_tuning_config.yaml:57]: try grid to provide a comprehensive search

File: examples/tuning/cta_svm/results/pipeline/best_test_acc.csv [hunk 13]
```diff
@@ -0,0 +1,4 @@
+id,acc,_step,_wandb_runtime,_runtime,test_acc,_timestamp,pipeline.0,pipeline.1,pipeline.2
```
Comment (xingzhongyu) [inline examples/tuning/cta_svm/results/pipeline/best_test_acc.csv:1]: Among all pipelines (missing allowed), the three configurations with the highest test_acc

File: examples/tuning/cta_svm/results/params/1st_best_test_acc.csv [hunk 14]
```diff
@@ -0,0 +1,2 @@
+id,_runtime,_step,_timestamp,_wandb_runtime,acc,params.0.max_cells,params.0.max_counts,params.0.min_cells,params.0.min_counts,params.0.order,params.1.n_components,test_acc
```
Comment (xingzhongyu) [inline examples/tuning/cta_svm/results/params/1st_best_test_acc.csv:1]: get result by save_summary_data test_acc is 0.95

File: examples/tuning/cta_svm/results/params/2nd_best_test_acc.csv [hunk 15]
```diff
@@ -0,0 +1,2 @@
+id,_runtime,_step,_timestamp,_wandb_runtime,acc,params.0.flavor,params.0.max_disp,params.0.max_mean,params.0.min_disp,params.0.min_mean,params.0.n_bins,params.1.n_components,test_acc
```
Comment (xingzhongyu) [inline examples/tuning/cta_svm/results/params/2nd_best_test_acc.csv:1]: test_acc is 0.962

File: examples/tuning/cta_svm/results/params/3rd_best_test_acc.csv [hunk 16]
```diff
@@ -0,0 +1,2 @@
+id,_runtime,_step,_timestamp,_wandb_runtime,acc,params.0.max_cells,params.0.max_counts,params.0.min_cells,params.0.min_counts,params.0.order,params.1.n_components,test_acc
```
Comment (xingzhongyu) [inline examples/tuning/cta_svm/results/params/3rd_best_test_acc.csv:1]: test_acc is 0.947

File: examples/tuning/cta_svm/pipline_run.sh (hunk 9 above)
Comment (xingzhongyu) [inline examples/tuning/cta_svm/pipline_run.sh]: run all subsets to get comprehensive results.

File: examples/tuning/cta_svm/config_yamls/params/1st_test_acc_params_tuning_config.yaml [hunk 17]
```diff
@@ -0,0 +1,65 @@
+type: preprocessor
```
Comment (xingzhongyu) [inline examples/tuning/cta_svm/config_yamls/params/1st_test_acc_params_tuning_config.yaml:1]: Under the highest configuration of test_acc in step2, the parameter range in step3

File: examples/tuning/cta_svm/pipline_run.sh [hunk 18]
```diff
@@ -0,0 +1,8 @@
+# python examples/tuning/cta_svm/main.py --config_dir=config_yamls/pipeline/subset_0_  # don't need 4 7 4 28 16 28
```
Comment (xingzhongyu) [inline examples/tuning/cta_svm/pipline_run.sh:1]: run all subsets to get comprehensive results.

File: examples/tuning/cta_svm/config_yamls/params/2nd_test_acc_params_tuning_config.yaml [hunk 19]
```diff
@@ -0,0 +1,65 @@
+type: preprocessor
```
Comment (xingzhongyu) [inline examples/tuning/cta_svm/config_yamls/params/2nd_test_acc_params_tuning_config.yaml:1]: Under the second highest configuration of test_acc in step2, the parameter range in step3

File: dance/pipeline.py [hunk 20]
```diff
@@ -848,15 +852,83 @@ def flatten_dict(d, *, parent_key="", sep="_"):
...
+        subset yaml files save path
+    file_path
+        python execution file, usually the file name is main
+    log_dir
+        log storage path
+    required_indexes
+        index in required type of origin yaml
+    Returns
+    --------
+    command_str: str
```
Comment (xingzhongyu) [inline dance/pipeline.py:914]: Returns the command for execution

File: examples/tuning/cta_svm/pipeline_run.sh [hunk 21]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
```
Comment (xingzhongyu) [inline examples/tuning/cta_svm/pipeline_run.sh:1]: run all subsets to get comprehensive results.

File: dance/pipeline.py [hunk 22]
```diff
@@ -848,15 +853,90 @@ def flatten_dict(d, *, parent_key="", sep="_"):
...
+    command_str = "#!/bin/bash \nlog_dir=" + log_dir + " \nmkdir -p ${log_dir}\n"
+    for index, subset in enumerate(subsets):
+        config_copy = dict_config.copy()
+        config_copy[tune_mode] = subset
+        configs.append(config_copy)
+        save_path = f"{save_directory}/subset_{index}_{tune_mode}_tuning_config.yaml"
+        if save_config:
+            OmegaConf.save(config_copy, save_path)
+        count = reduce(mul, [len(p["include"]) if "include" in p else 1 for p in subset])
+        config_dir = os.path.relpath(os.path.dirname(save_path), os.path.dirname(os.path.join(root_path, file_path)))
```
Comment (xingzhongyu) [inline dance/pipeline.py:940]: Get the relative path between the file execution path and the configuration file path

File: dance/pipeline.py [hunk 23]
```diff
@@ -848,15 +853,137 @@ def flatten_dict(d, *, parent_key="", sep="_"):
...
+    count = 0
+    for row in pipeline_names:
+        pipeline = []
+        row = [i for i in row]
+        for i, f in zip(required_indexes, required_funs):
+            row.insert(i, f)
+        for x in row:
+            for k in conf.pipeline:
+                if k["target"] == x:
+                    pipeline.append(k)
```
Comment (xingzhongyu) [inline dance/pipeline.py:985]: The corresponding pipeline was found from the default parameter search range file

File: dance/pipeline.py [hunk 24]
```diff
@@ -848,15 +853,137 @@ def flatten_dict(d, *, parent_key="", sep="_"):
...
+    pipeline_names = result.loc[:, columns].values
+    count = 0
+    for row in pipeline_names:
+        pipeline = []
+        row = [i for i in row]
+        for i, f in zip(required_indexes, required_funs):
+            row.insert(i, f)
+        for x in row:
+            for k in conf.pipeline:
+                if k["target"] == x:
```
Comment (xingzhongyu) [inline dance/pipeline.py:984]: The corresponding pipeline was found from the default parameter search range file

File: dance/pipeline.py [hunk 25]
```diff
@@ -848,15 +853,137 @@ def flatten_dict(d, *, parent_key="", sep="_"):
...
+    conf = OmegaConf.load(conf_load_path)
+    result = pd.read_csv(result_load_path)
+    columns = sorted([col for col in result.columns if col.startswith("pipeline")])
+    pipeline_names = result.loc[:, columns].values
+    count = 0
+    for row in pipeline_names:
+        pipeline = []
+        row = [i for i in row]
+        for i, f in zip(required_indexes, required_funs):
+            row.insert(i, f)
```
Comment (xingzhongyu) [inline dance/pipeline.py:981]: Insert required functions

File: dance/pipeline.py [hunk 26]
```diff
@@ -848,15 +853,137 @@ def flatten_dict(d, *, parent_key="", sep="_"):
...
+    """
+    DANCEDIR = Path(__file__).resolve().parent.parent
+    root_path = DANCEDIR if root_path is None else root_path
+    conf_save_path = os.path.join(root_path, conf_save_path)
+    conf_load_path = os.path.join(root_path, conf_load_path)
+    result_load_path = os.path.join(root_path, result_load_path)
+    conf = OmegaConf.load(conf_load_path)
+    result = pd.read_csv(result_load_path)
+    columns = sorted([col for col in result.columns if col.startswith("pipeline")])
+    pipeline_names = result.loc[:, columns].values
```
Comment (xingzhongyu) [inline dance/pipeline.py:975]: Get the pipeline name in the results

File: dance/transforms/filter.py [hunk 27]
```diff
@@ -1053,6 +1053,17 @@ def __init__(self, layer: Optional[str] = None, n_top_genes: Optional[int] = 100
                          flavor=flavor, subset=subset, inplace=inplace, batch_key=batch_key, **kwargs)
 
 
+@register_preprocessor("filter", "gene")
+class FilterGenesPlaceHolder(BaseTransform):
+    """Used as a placeholder to skip the process."""
```
Comment (xingzhongyu) [inline dance/transforms/filter.py:1058]: Used as a placeholder to skip the process.

File: examples/tuning/cta_svm/pipeline_tuning_config.yaml [hunk 28]
```diff
@@ -1,17 +1,46 @@
 type: preprocessor
 tune_mode: pipeline
 pipeline:
+  - type: normalize
+    include:
+      - ScaleFeature
+      - ScTransform
+      - Log1P
+      - NormalizeTotal
+      - NormalizePlaceHolder
```
Comment (xingzhongyu) [inline examples/tuning/cta_svm/pipeline_tuning_config.yaml:10]: add PlaceHolder to skip the preprocess

File: examples/tuning/cta_svm/placeholder_pipeline_tuning_config.yaml [hunk 29]
```diff
@@ -0,0 +1,63 @@
+type: preprocessor
```
Comment (xingzhongyu) [inline examples/tuning/cta_svm/placeholder_pipeline_tuning_config.yaml:1]: add PlaceHolder to skip the preprocess

File: dance/transforms/filter.py (hunk 27 above)
Comment (RemyLau) [reply dance/transforms/filter.py:1058]: This can stay for now, but will need to be removed eventually.

I will implement a more general bypass mechanism from the pipeline object so that there's always an option to bypass a specific step.

File: dance/transforms/normalize.py (hunk 4 above)
Comment (RemyLau) [reply dance/transforms/normalize.py:260]: Ok fair enough, I'll also keep thinking about it

**Review by RemyLau — APPROVED** [review]
